### DISCUSSION TITLE
0x8FB - 0x9EB Request Values, latest FPA Template

### DISCUSSION THREAD

**Discussion by Cuzoe** [jilleb/MQB-FPA#73] [Exploration and research]
First... I know there is an older discussion about the 8FB range, but I thought this was different enough for a new thread.

So I had modified my dataset some time ago and been pretty happy with it.  Then saw the template has been updated so I thought why not play around with that.  I reading up on the [wiki here](https://github.com/jilleb/MQB-FPA/wiki/Dataset#0x8fb---0x9eb-request-values) about the request values.  There are 15 columns, which correspond, it says, to driving modes.  In the link above HVAC is used as an example, with a value for Normal (2) and Eco (5).  This I understand.

My understanding from this is that when driving profile is selected to Normal, HVAC is sent (2) for Normal mode.  And this value should show up in the (HVAC) module's real time values for "requested" and "actual" value, assuming the module accepts the request.

But this logic does not work when looking at a dataset configured with the DCC slider.  In one of these datasets (I'm using the example one from the main page, but my personal dataset is the same) this 8FB Range Ttable shows values from 1-15, in columns 1-15.  If the same HVAC request logic holds true for DCC... then when driving profile is selected to Normal, DCC is sent (2) for Normal mode.  And this value should show up in the (DCC) module's real time values for "requested" and "actual" value.  But this is not the case!

In my car... I select driving profile Normal, while watching DCC module real time request/actual.  The request/actual show "8," which is correct based on 0x386h, which has Normal profile send (8) to the DCC module.  In this case the DCC module expecting values from 1-15.  I did verify in the DCC request/actual that my profile is sending request correctly...
```
Eco - DCC receives value 8 (8)
Comfort - DCC receives value 3 (3)
Normal - DCC receives value 8 (8)
Sport - DCC receives value 13 (d)
Race - DCC receives value 15 (f) [I don't have this FEC, but it is configured]
```

So using the 8FB Range Table does not, to my understanding, actually represent the values being sent to the module for a given mode, as suggested by the column names.  Rather, it seems to identify the values that are _possible_ to send.  Admittedly, those line up with driving profiles in many cases but that is not absolute, as is the case with DCC.

Incidentally, having this table has still helped me, as I could not figure out why I was seeing Eco+ for background lighting with my driving profile set to Individual... but **only** when I reset driving profiles.  Once I selected anything else from the drop down I could never re-select Eco+ without doing a reset.  My suspicion was/is that this erroneous (not properly configured) value request is preventing me from being able to set my ambient lighting color mode to "coupled/automatic" vs "manual" color selection.  Coupled meaning based on driving mode.  But that's another topic, kind of, haha.

**Comment by Cuzoe** [comment]
Well, I figured out why Eco+ was showing when I reset driving modes.  Silly mistake on my part.  Sending (7) at 0x420h, which the template says _Individual/Eco+_.  But then in 0x661h I was sending (46 = 2+4+8+32) which gives the options for _Comfort, Normal, Sport and Eco_.
- I suppose I could set 0x661h to 174 (2+4+8+32+128) which would give selectable options _Comfort, Normal, Sport, Eco and Eco+_.
- But then I don't want to assign so many colors, prefer to use color slider on the Ambient Light settings infotainment screen.

I guess my understanding of the 8FB range is "evolving," haha.  And it will take lots of flashing then checking real time values to confirm everything is doing what I would expect/want but I think the following general rules are safe... what do you all think?

For the modules that accept values 1-5 this seems to also correspond to Comfort, Normal, Sport, Offroad, Eco
- In the datasets that have Race instead of Sport, individual (non-grouped) controls seem to get value 3, which makes sense, as the Race mode tends to "replace" Sport mode
- And also if you look at Unknown 38x research in the template, I see that the grouped controls for a profile with Race is still sending value 3 for the Race profile... which again makes sense with Race mode "replacing" Sport mode

Then there are modules like DCC with software index C, that will accept values from 1-15
- These no longer correspond to Comfort, Normal, Sport, Offroad, Eco, etc.

And then there is background lighting, which effectively accepts values from 0-15 (which is ultimately sent to the BCM)
- These correspond to (facelift) BCM adaptation _Ambientelicht Zuordnung der Farbe zum Fahrprofil-Fahrprofil_0_ through _Fahrprofile_15_
- Translation = _Ambient light assignment of the color to the driving profile-driving profile_0_ through _profile_15_
- Pretty neat stuff, the values set in this list of adaptations correspond to your RGB Color List
  - This could be anything but typically you see Normal = Blue, Eco = Green, Sport = Red, etc.

**Reply by olli991** [reply]
If you look into the wiki the understanding of the available options is quite easy.

Switch the view to Binary and then every 1 stands for an activated option. You'll always have 15 bit there. Depending on which 1 you set the according mode will be shown in the dropdown:
<html>
<body>


15 | 14 | 13 | 12 | 11 | 10 | 9 | 8 | 7 | 6 | 5 | 4 | 3 | 2 | 1 | 0
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
? | ? | Sand | ? | ? | Snow | Off | On | Eco+ | Race | Eco | Offroad | Sport | Normal | Comfort | ?


</body>
</html>

**Reply by Cuzoe** [reply]
I understand it now.  The only part that was confusing (to me at first, and seems to be the confusion that @thaibq had) is that the Modes don't apply for DCC index C.  For example, setting 13 for the DCC module is not "Sand" mode/profile.  It is 2 clicks away from full firm, and corresponds to "Sport".
- Where on Index B "Sport" mode was setting 3.

_Edit: And I understand that in the 365h range, setting Sport mode to send "13" to the DCC module still shows "Sand" in the template, but on the screen (and in DCC function) this is Sport._

Basically what I was/am saying is the profile (Comfort, Normal, Sport, Offroad, Eco, Race, etc.) names do not _always_ apply, although they will be correct 95% of the time.

**Reply by olli991** [reply]
Yes in DCC those names obviously do not apply...

**Reply by Cuzoe** [reply]
Haha, yes it is obvious **now**.  But looking at the wiki and then looking at a DCC slider enabled dataset could leave someone (me) confused, for a moment.

I have a "more important" question that maybe you can answer while you're here...
I have RGB ambient lighting in my car (also infotainment and dash change colors).  I have the colors assigned in the fpa dataset, and have them assigned in the BCM adaptations... all that works when switching modes.
- What does not work, is that I don't have the option to choose Automatic or Manual on the Ambient Lighting page
![Capture](https://user-images.githubusercontent.com/101527335/190510947-b9310065-558d-4d80-9484-f686543191b6.JPG)
- Sorry for the crappy, not in English picture but that is the selection I want.
- Instead I can only choose "Off," even though my BCM adaptations are set properly for Auto/Manual
  - These are the FPA coupling BCM adaptations
 ```
Ambiente_Farbwahl_FPA_waehlbare_Kopplung - active
Ambiente_Farbwahl_FPA_waehlbare_Kopplung_Status_hmi_default - coupled
Translation...
Ambiance_colour_choice_FPA_selectable_coupling - active
Ambience_color_choice_FPA_selectable_coupling_status_hmi_default - coupled
```

When I change driving modes the colors change as assigned, good.
When I'm Individual I can use the slider to pick a color, which is saved and color returns to that when switching back to Individual, good.
But the Manual/Automatic option should let me do the following:
- Set driving mode to Sport (lights turn red)
- Go to Ambient Lighting page and select _manual_, and change the color while still in Sport
- But then later select _automatic_ to re-couple Sport with the red color

**Reply by olli991** [reply]
Coding is correct, mine isn't different. Do you also coded DTC 19: Long coding (Text mode) - FPA_Funktion_AMB -> Active (Byte 11 Bit 7) to make activate it in the FPA... maybe those are realated, i dunno.

**Reply by Cuzoe** [reply]
Yup, I have that coded.  I _think_ when I didn't have that coded the color didn't change at all when switching between driving modes.  It's kind of strange.  Assuming you have the Auto/Manual option, would you mind sharing your FPA dataset so I can compare and see if my this is where I have a problem.

This is the dataset I'm going to flash this afternoon, with a few changes I've made after looking at some of the Original Profiles available here.  I'll report back of course, but that will not be several hours.
[Ambient Test.zip](https://github.com/jilleb/MQB-FPA/files/9579288/Ambient.Test.zip)

**Comment by thaibq** [comment]
Am i right to understand that 8FB range is only relating to selection in INDIVIDUAL mode? (not Driving mode button)
Example: DCC in individual mode has Comfort, Normal, Sport -> it requests value 1,2,3
Or if having DCC slider in Individual -> It requests value 1-15

And 365h range is exactly what relates to selection of Driving mode button.

**Reply by Cuzoe** [reply]
That is not exactly correct, 8FB range is always used, Individual and Drive mode button (365h range)

### DCC module with software A/B expects values 1-3
```
Value 1 = Comfort
Value 2 = Normal
Value 3 = Sport
```
If you have DCC software A/B you cannot use the slider, and 8FB should have 1-3
- 365h range, you set values 1-3 for Driving Mode button, set each profile to what you want
  - Normally this would be Comfort, Normal Sport > 1, 2, 3 as you mention



### DCC module with software C expects values  1-15 
```
Value 1 = Softer than Comfort
Value 3 = Comfort (On DCC software A/B, equivalent to Value 1) 
Value 8 = Middle (On DCC software A/B, equivalent to Value 2)
Value 13 = Sport (On DCC software A/B, equivalent to Value 3)
Value 15 = Firmer than Sport
```
If you have DCC software C, that means you can use the slider for individual, and 8FB should have 1-15
- 365h range, you can set values from 1-15 for Driving mode button, set each profile to what you want
  - For example, you could set Eco to send value 1 (Softer than Comfort)
  - For another example, you could set Sport to send value f (Firmer than Sport), if you want
- If you set Comfort, Normal, Sport > 1, 2, 3 then all of your modes will be soft (see the table above)

**Reply by thaibq** [reply]
@Cuzoe: Thank you.
So, 8FB is to define value of each profile, then 365F is to call each profile with its value defined in 8FB, then send such value to ECU.

Example with DCC:
a) if 8FB defines Comfort = value 1, then 365F calls DCC = Comfort -> When we press "Comfort button", It will send value 1 to DCC ECU -> Car will work with Comfort mode // This is correct setting
b) if 8FB defines Comfort value = 2, then 365F calls DCC = Comfort -> When we press "Comfort button", It will send value 2 to DCC ECU -> Car will work with Normal mode // this is incorrect setting, i just mentioned to express my thinking

Do you agree?

**Reply by Cuzoe** [reply]
Yes, you are correct.  The 365 range will send the values specified in 8FB, even if they are incorrect.
- Incorrect can mean wrong mode like your example, Normal instead of Comfort
- Incorrect can also mean a value the module does not accept
  - module just ignores the value, driver may or may not notice, haha

**Reply by thaibq** [reply]
@Cuzoe: Thank you. I figured it out ^^
Do you have Race button in your HMI screen? I am trying to enable it because i guest it doesn't need FEC.
And one more thing: My car (Tiguan Allspace) has Knob button to select profile. When i check Swap status it showed that SNOW fec is invalid. But the car still works with snow mode by Knob button. So i think there must be a way to active Comfort without FEC (like my snow mode above). Just trying to investigate...

**Reply by thaibq** [reply]
@jilleb 
![image](https://user-images.githubusercontent.com/113581081/190552423-9de46081-d79b-4d0c-9fb3-b58d1103fcb4.png)
I have this knob to active snow mode, while i check swap status of snow, it is invalid.
It means that snow mode can work without Fec. How to you think?

**Reply by olli991** [reply]
If you go into snow do you have an animation or info for that on the HMI? 
To my knowledge fecs are only needed to show the dedicated button in HMI. 

Via options in Individual you can go into every mode for example no matter what fecs you have.

**Reply by thaibq** [reply]
@olli991: Of couse, HMI showed animation and info of snow mode:
![image](https://user-images.githubusercontent.com/113581081/190581809-c8e227aa-1e6b-4841-b66c-f90d408661fd.png)

Activated by Knob:
![image](https://user-images.githubusercontent.com/113581081/190582031-1b7b59c1-026c-451a-8e59-95083aff8ebb.png)

Swap said invalid:
![image](https://user-images.githubusercontent.com/113581081/190582138-120a1221-c747-450e-a8f3-38b77bbcf0de.png)

Is this strange?

**Reply by olli991** [reply]
I would say it is normal because you only get the animation. You can't change into snow a button in HMI

**Reply by thaibq** [reply]
@olli991 Thank you, i got this point. It is like making eco button working in Comfort mode.
So i can also use Knob to switch to Comfort mode as quick action only, No way to display comfort button in HMI without Fec

**Reply by Cuzoe** [reply]
That is interesting.  What shows if you select the "I" for info on that Snow mode HMI page?  If you see what is there you can re-create Snow mode.

I think jilleb activated all the HMI buttons I'm his car...
https://www.drive2.com/l/527768880941106936/
Without the FEC the modules may not have "real" Snow/Sand/Offroad mode (8FB range).  But you could have the "faked" modes send values (365 range) modes you do have.
- For example I could enable Race HMI, then set all the same values (365 range) as Sport.

**Reply by thaibq** [reply]
@Cuzoe Yes, yes... i am thinking of that. There is one slot in Profile [5] still = 00 not set. I will try adding Race there to see what happen :)

**Reply by thaibq** [reply]
@Cuzoe Not successful to active race button in HMI haha. Only Fec/swap can help

**Reply by Cuzoe** [reply]
Interesting, though not unexpected.  But you also don't have Snow FEC, correct?

Can you only get to Snow with the rotary switch?  Not with the mode button?

Can you attach the data set you loaded with the race profile that does not show?  I'll be curious to see what's going on in there, haha.

**Reply by thaibq** [reply]
"Can you only get to Snow with the rotary switch? Not with the mode button?" -> Yes, exactly.
For Race, i just tried simply this modification (But not successful :D )
![image](https://user-images.githubusercontent.com/113581081/190943596-3dd7a948-314d-493d-afab-02d9432850ea.png)
